packages/gatsby-core-utils/src/fetch-remote-file.ts

```diff
@@ -44,6 +48,190 @@ const INCOMPLETE_RETRY_LIMIT = process.env.GATSBY_INCOMPLETE_RETRY_LIMIT
   ? parseInt(process.env.GATSBY_INCOMPLETE_RETRY_LIMIT, 10)
   : 3
 
+let fetchCache = new Map()
+let latestBuildId = ``
+
+export async function fetchRemoteFile(
+  args: IFetchRemoteFileOptions
+): Promise<string> {
+  const BUILD_ID = global.__GATSBY?.buildId ?? ``
+  if (BUILD_ID !== latestBuildId) {
+    latestBuildId = BUILD_ID
+    fetchCache = new Map()
+  }
+
+  // If we are already fetching the file, return the unresolved promise
+  const inFlight = fetchCache.get(args.url)
+  if (inFlight) {
+    return inFlight
+  }
+
+  // Create file fetch promise and store it into cache
+  const fetchPromise = fetchFile(args)
+  fetchCache.set(args.url, fetchPromise)
+
+  return fetchPromise.catch(err => {
+    fetchCache.delete(args.url)
+
+    throw err
+  })
+}
+
+function pollUntilComplete(
+  cache: GatsbyCache,
+  url: string,
+  buildId: string,
+  cb: (result: string | null) => void
+): void {
+  if (!IS_WORKER) {
+    // We are not in a worker, so we shouldn't use the cache
+    return void cb(null)
+  }
+
+  cache.get(cacheIdForWorkers(url)).then(entry => {
+    if (!entry || entry.buildId !== buildId) {
+      return void cb(null)
+    }
+
+    if (entry.status === `complete`) {
+      cb(entry.result)
+    } else {
+      setTimeout(() => {
+        pollUntilComplete(cache, url, buildId, cb)
+        // Magic number
+      }, 500)
+    }
```

Comment:
What happens if a worker starts polling but then the "main" worker fails (i.e. `requestRemoteNode` throws)? I don't see how it can exist from this function in this case. Can we have a test case for it as it is a typical scenario?

Reply:
We keep on hanging 😂 good point but wouldn't the process fail anyway?

Reply:
Depends on how folks use it. If it happens in custom resolver - the corresponding field can just return `null` (graphql allows partial errors for nullable fields)